src/box/lua/upgrade.lua

```diff
@@ -1357,6 +1360,8 @@ local function upgrade(options)
     options = options or {}
     setmetatable(options, {__index = {auto = false}})
 
+    ffi.C.box_dd_operations_allowed = true
```

Comment:
You allow ddl operations even before running the upgrade handlers.
If any of the handlers fail, we are left with old schema and without ddl protection.

Let's either restore the original value of `box_dd_operations_allowed` upon failure, or (better IMO) let's take a slightly different approach, as proposed in my comment.

Reply:
Discussed f2f and decided to forbid the following operations in Lua:
- box.schema.space.create
- box.schema.space.drop
- box.schema.space.alter
- box.schema.index.create
- box.schema.index.drop
- box.schema.index.alter
- box.schema.sequence.create
- box.schema.sequence.drop
- box.schema.sequence.alter
- box.schema.func.create
- box.schema.func.drop
- box.internal.collation.create
- box.internal.collation.drop